File: backend/app/routers/ai.py
from typing import Dict, Any
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database.database import get_db
from app.core.dependencies import get_current_user
from app.database.models import User, Application, JobRequirement, VerificationType
from app.schemas.ai import (
    ApplicationAnalysisResponse, EvidenceGraphResponse,
    ProveRequirementResponse, VerificationChallengeRequest,
    VerificationChallengeResponse, VerificationInterviewQuestionRequest,
    VerificationInterviewQuestionResponse, CandidateAnswerEvaluationRequest,
    CandidateAnswerEvaluationResponse, FinalAnalysisResponse
)
from app.services.application_service import application_service
from app.services.verification_service import verification_service
from app.ai.final_analyzer import final_analyzer

router = APIRouter(prefix="/ai", tags=["AI Reasoning & Verification"])

@router.post(
    "/applications/{application_id}/analyze",
    response_model=ApplicationAnalysisResponse,
    summary="Run AI Match & Evidence Analysis",
    description="Analyzes candidate resume against vacancy requirements, extracts concrete evidence snippets, and computes match statuses."
)
async def analyze_application(
    application_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    try:
        return await application_service.run_ai_analysis_pipeline(application_id, db)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={
                "error": "AI_ANALYSIS_FAILED",
                "message": f"Unable to complete AI analysis at this time: {str(e)}"
            }
        )

@router.get(
    "/applications/{application_id}/evidence-graph",
    response_model=EvidenceGraphResponse,
    summary="Get Evidence Graph for Application",
    description="Returns graph nodes and edges representing: Requirement -> Evidence -> Reasoning -> Decision."
)
def get_evidence_graph(
    application_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    app = db.query(Application).filter(Application.id == application_id).first()
    if not app:
        raise HTTPException(status_code=404, detail="Application not found.")

    return application_service.generate_evidence_graph(app)

@router.post(
    "/applications/{application_id}/requirements/{requirement_id}/prove",
    response_model=ProveRequirementResponse,
    summary="Determine Verification Method ('Prove This Skill')",
    description="Analyzes the requirement and current match status to suggest whether an AI Challenge, Interview, or Document Upload is required."
)
def prove_requirement(
    application_id: int,
    requirement_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    return verification_service.get_proof_recommendation(
        application_id=application_id,
        requirement_id=requirement_id,
        db=db
    )

@router.post(
    "/verification/challenge",
    response_model=VerificationChallengeResponse,
    summary="Generate AI Verification Challenge",
    description="Generates a targeted, scenario-based technical challenge specifically for the requirement."
)
async def generate_challenge(
    payload: VerificationChallengeRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    return await verification_service.create_challenge(
        application_id=payload.application_id,
        requirement_id=payload.requirement_id,
        db=db
    )

@router.post(
    "/verification/interview-question",
    response_model=VerificationInterviewQuestionResponse,
    summary="Generate AI Interview Question",
    description="Generates a targeted technical interview question assessing practical experience with the requirement."
)
async def generate_interview_question(
    payload: VerificationInterviewQuestionRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    return await verification_service.create_interview_question(
        application_id=payload.application_id,
        requirement_id=payload.requirement_id,
        db=db
    )

@router.post(
    "/verification/evaluate",
    response_model=CandidateAnswerEvaluationResponse,
    summary="Evaluate Candidate Answer & Update Status",
    description="Evaluates technical depth and evidence. If STRONG_EVIDENCE is proven, updates status to VERIFIED while saving full audit history."
)
async def evaluate_candidate_answer(
    payload: CandidateAnswerEvaluationRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    return await verification_service.evaluate_candidate_answer(
        application_id=payload.application_id,
        requirement_id=payload.requirement_id,
        question=payload.question,
        candidate_answer=payload.candidate_answer,
        verification_type=VerificationType.AI_INTERVIEW,
        db=db
    )

@router.get(
    "/applications/{application_id}/final-analysis",
    response_model=FinalAnalysisResponse,
    summary="Get Final Comprehensive Analysis & Audit Report",
    description="Returns detailed breakdown, overall requirement coverage, verified evidence, verification history, and human overrides."
)
def get_final_analysis(
    application_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    app = db.query(Application).filter(Application.id == application_id).first()
    if not app:
        raise HTTPException(status_code=404, detail="Application not found.")

    return final_analyzer.generate_final_report(app)

File: frontend/src/api/ai.ts
import { apiClient } from './client';
import type { RequirementMatchItem, EvidenceResponseItem } from './applications';

export interface ApplicationAnalysisResponse {
  application_id: number;
  overall_score: number;
  analyzed_at: string;
  requirement_matches: RequirementMatchItem[];
}

export interface GraphNode {
  id: string;
  type: string;
  label: string;
  metadata?: Record<string, any>;
}

export interface GraphEdge {
  source: string;
  target: string;
  label: string;
}

export interface EvidenceGraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface FinalAnalysisResponse {
  application_id: number;
  candidate_name: string;
  vacancy_title: string;
  company_name: string;
  overall_score: number;
  match_status_summary: {
    verified: number;
    partial: number;
    unverified: number;
    gap: number;
  };
  requirement_matches: RequirementMatchItem[];
  evidence_graph: EvidenceGraphResponse;
  verification_history: any[];
  hr_challenges: any[];
  final_recommendation: string;
  audit_defense_available: boolean;
}

export interface ProveRequirementResponse {
  requirement_id: number;
  requirement_text: string;
  current_status: string;
  recommended_method: 'AI_CHALLENGE' | 'AI_INTERVIEW' | 'DOCUMENT_UPLOAD';
  rationale: string;
}

export interface VerificationChallengeRequest {
  application_id: number;
  requirement_id: number;
}

export interface VerificationChallengeResponse {
  challenge_id: string;
  application_id: number;
  requirement_id: number;
  challenge_type: string;
  title: string;
  problem_statement: string;
  starter_code?: string | null;
  evaluation_criteria: string[];
}

export interface VerificationInterviewQuestionRequest {
  application_id: number;
  requirement_id: number;
}

export interface VerificationInterviewQuestionResponse {
  question_id: string;
  application_id: number;
  requirement_id: number;
  question: string;
  context: string;
  suggested_focus_areas: string[];
}

export interface CandidateAnswerEvaluationRequest {
  application_id: number;
  requirement_id: number;
  question: string;
  candidate_answer: string;
}

export interface CandidateAnswerEvaluationResponse {
  application_id: number;
  requirement_id: number;
  technical_depth_score: number;
  evidence_evaluation: 'STRONG_EVIDENCE' | 'PARTIAL_EVIDENCE' | 'INSUFFICIENT_EVIDENCE';
  reasoning: string;
  new_match_status: string;
  status_updated: boolean;
}

export const aiApi = {
  async analyzeApplication(applicationId: number): Promise<ApplicationAnalysisResponse> {
    return apiClient<ApplicationAnalysisResponse>(`/ai/applications/${applicationId}/analyze`, {
      method: 'POST',
    });
  },

  async getEvidenceGraph(applicationId: number): Promise<EvidenceGraphResponse> {
    return apiClient<EvidenceGraphResponse>(`/ai/applications/${applicationId}/evidence-graph`);
  },

  async getFinalAnalysis(applicationId: number): Promise<FinalAnalysisResponse> {
    return apiClient<FinalAnalysisResponse>(`/ai/applications/${applicationId}/final-analysis`);
  },

  async getRequirements(applicationId: number): Promise<any[]> {
    return apiClient<any[]>(`/ai/requirements/${applicationId}`);
  },

  async getEvidence(applicationId: number): Promise<EvidenceResponseItem[]> {
    return apiClient<EvidenceResponseItem[]>(`/ai/evidence/${applicationId}`);
  },

  async proveRequirement(applicationId: number, requirementId: number): Promise<ProveRequirementResponse> {
    return apiClient<ProveRequirementResponse>(`/ai/applications/${applicationId}/requirements/${requirementId}/prove`, {
      method: 'POST',
    });
  },

  async generateChallenge(payload: VerificationChallengeRequest): Promise<VerificationChallengeResponse> {
    return apiClient<VerificationChallengeResponse>('/ai/verification/challenge', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },

  async generateInterviewQuestion(payload: VerificationInterviewQuestionRequest): Promise<VerificationInterviewQuestionResponse> {
    return apiClient<VerificationInterviewQuestionResponse>('/ai/verification/interview-question', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },

  async evaluateAnswer(payload: CandidateAnswerEvaluationRequest): Promise<CandidateAnswerEvaluationResponse> {
    return apiClient<CandidateAnswerEvaluationResponse>('/ai/verification/evaluate', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },
};
